### ISSUE TITLE
Different result between v8.1.2 and v8.2 on same dataset

### ISSUE THREAD

**Issue by a773783082** [ultralytics/ultralytics#13871]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I trained yolov8n on my dataset in version 8.1.2 and 8.2, epochs=90, close_moasic=20. After training, I found the val and test result is much different between two kinds of version.
Also,  in "results.png", train loss at the stage of close_moasic is increased in v8.2, relatively it is decreased in v8.1.2.
I want to know what is different between these different version codes, and how cat I get same results in v8.2?
Thank you!!

v8.1.2 results.png
![Snipaste_2024-06-22_10-40-21](https://github.com/ultralytics/ultralytics/assets/57038246/ebc08c1e-fed7-4883-8c47-30a5e0cdb217)

v8.2 results.png
![Snipaste_2024-06-22_10-40-29](https://github.com/ultralytics/ultralytics/assets/57038246/af083d2c-7594-4b07-8989-fc9462e0500c)

**Comment by nowbug** [comment]
I have the same issue, and it turns out that it's not specifically 8.2; there are noticeable differences even between versions 8.1.0 and 8.1.47 on the same dataset. The conclusion I've drawn is that the training map accuracy with version 8.1.0 is higher than that of versions 8.1.47 and 8.2.x. Although I'm unaware of the precise reason behind this, I share your confusion. I've been compelled to downgrade to 8.1.0 in order to regain the level of precision required for this part of my work.

**Comment by glenn-jocher** [comment]
Hello @nowbug,

Thank you for bringing this to our attention. The differences you're observing between YOLOv8 versions are likely due to updates and optimizations made to the codebase, which can affect training dynamics and results.

To help us investigate this further, could you please provide a minimum reproducible example of your training setup? This includes your dataset, configuration files, and any specific training commands you used. You can refer to our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details on what to include. This will allow us to reproduce the issue on our end and identify any potential causes.

Additionally, please ensure that you are using the latest versions of `torch` and `ultralytics` packages. You can upgrade them using the following commands:
```bash
pip install --upgrade torch ultralytics
```

In the meantime, here are a few steps you can take to potentially mitigate the differences:
1. **Check Hyperparameters:** Ensure that all hyperparameters, including learning rate, batch size, and augmentation settings, are consistent across different versions.
2. **Seed Initialization:** Set a fixed random seed to ensure reproducibility of results across different runs and versions.
3. **Review Changelog:** Check the [Ultralytics YOLOv8 changelog](https://github.com/ultralytics/ultralytics/releases) for any updates or changes that might affect training behavior.

We appreciate your patience and cooperation as we work to resolve this issue. If you have any further questions or need additional assistance, please feel free to ask.

**Comment by a773783082** [comment]
Thanks for your help! 
The code on remote server is not available for me now and the dataset is closed source, I can only provide some hyperparameters for you.
```
torch: version 1.11 with cuda 113
```
In addition, I trained my models on download code, with activation of `Conv` from `SiLU` to `ReLU`. And optimizer's paramemter is `auto` to set `SGD optim`.

**Comment by glenn-jocher** [comment]
Hi @a773783082,

Thank you for the additional details! To help us investigate further, could you please provide a minimum reproducible example when you have access to your code? This is crucial for us to reproduce and diagnose the issue effectively. You can refer to our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more information.

In the meantime, please ensure you're using the latest versions of `torch` and `ultralytics`:
```bash
pip install --upgrade torch ultralytics
```

Given your setup, here are a few things to check:
1. **Hyperparameters Consistency:** Ensure all hyperparameters, including learning rate, batch size, and augmentation settings, are consistent across different versions.
2. **Seed Initialization:** Set a fixed random seed to ensure reproducibility:
    ```python
    import torch
    import random
    import numpy as np

    seed = 42
    torch.manual_seed(seed)
    np.random.seed(seed)
    random.seed(seed)
    ```

3. **Optimizer Settings:** Verify that the optimizer settings are correctly applied in both versions.

If you can share any additional details or code snippets, it would be greatly appreciated. We're here to help! 😊

**Comment by a773783082** [comment]
Hi @glenn-jocher ,

I retrained on v8.1.2 and 8.2, the results wera same as before. Here is my training code.
I replace activation of `Conv` from `SiLU` to `ReLU`, and set `wandb=False`, `wb=None`.

```python
from ultralytics import YOLO
from pathlib import Path
import shutil
import random
import torch
import numpy as np

import warnings
warnings.filterwarnings('ignore')

_seed = 584
torch.manual_seed(_seed)
np.random.seed(_seed)
random.seed(_seed)

## Python-3.9.1 torch-1.11.0+cu113 CUDA:0
## optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 57 weight(decay=0.0),  64 weight(decay=0.001), 63 bias(decay=0.0)

def train_detect(opt, **args):
    # load a pretrained model
    model = YOLO(model=opt.model_cfg, task="detect")
    if Path(opt.weights).exists():
        print("-"*20, "load pretrained weights")
        model.load(opt.weights)

    # Train the model
    model.train(
        project=opt.project,
        name=opt.name,
        data=opt.data,
        epochs=90,
        imgsz=416,
        batch=128,
        save_period=20,
        ## hyp
        cache=True,
        #seed=random.randint(1, 1024),   # seed=0
        multi_scale=False,
        close_mosaic=20,
        ## augmentations
        hsv_h=0.015,
        hsv_s=0.7,
        hsv_v=0.4,
        degrees=15,
        scale=0.5,
        translate=0.1,
        mosaic=1,
        mixup=0.,
    )
```

**Comment by glenn-jocher** [comment]
Hi @a773783082,

Thank you for providing your training code and additional details. This is very helpful! 😊

To further investigate the differences in results between YOLOv8 versions, it would be beneficial to have a minimum reproducible example. This allows us to replicate the issue on our end and identify any potential causes. You can refer to our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more information on what to include.

In the meantime, please ensure you are using the latest versions of `torch` and `ultralytics`:
```bash
pip install --upgrade torch ultralytics
```

Here are a few additional suggestions to help troubleshoot the issue:

1. **Hyperparameters Consistency:** Ensure all hyperparameters, including learning rate, batch size, and augmentation settings, are consistent across different versions.
2. **Seed Initialization:** You've already set a fixed random seed, which is great for ensuring reproducibility.
3. **Optimizer Settings:** Verify that the optimizer settings are correctly applied in both versions.

If you continue to experience discrepancies, please share any additional details or code snippets that might help us diagnose the issue further.

Thank you for your patience and cooperation. We're here to help! 😊
